pkg/build/util/jenkins.go

```diff
@@ -0,0 +1,164 @@
+// JenkinsPipelineStrategy template, used to instantiate the Jenkins service in
+// given namespace.
+type JenkinsPipelineTemplate struct {
+	Config          serverapi.JenkinsPipelineConfig
+	TargetNamespace string
+	kubeClient      *kclient.Client
+	osClient        *client.Client
+	items           []resourceMapping
+	ProcessErrors   []error
+	CreateErrors    []error
+}
+
+// Process processes the Jenkins template. If an error occurs
+func (t *JenkinsPipelineTemplate) Process() *JenkinsPipelineTemplate {
+	if len(t.items) > 0 {
+		return t
+	}
+	template, err := t.osClient.Templates(t.Config.Namespace).Get(t.Config.TemplateName)
+	if err != nil {
+		t.ProcessErrors = append(t.ProcessErrors, err)
+		return t
+	}
+	// TODO: All parameters must have defaults here. Should we allow setting
+	// parameters in build strategy?
+	pTemplate, err := t.osClient.TemplateConfigs(t.TargetNamespace).Create(template)
+	if err != nil {
+		t.ProcessErrors = append(t.ProcessErrors, fmt.Errorf("processing Jenkins template %s/%s failed: %v", t.Config.Namespace, t.Config.TemplateName, err))
+		return t
+	}
+	var mappingErrs []error
+	t.items, mappingErrs = mapJenkinsTemplateResources(pTemplate.Objects)
+	if len(mappingErrs) > 0 {
+		t.ProcessErrors = append(t.ProcessErrors, mappingErrs...)
+		return t
+	}
+	glog.V(4).Infof("Processed Jenkins pipeline template %s/%s", pTemplate.Namespace, pTemplate.Namespace)
+	return t
+}
+
+// Instantiate instantiates the Jenkins template in the target namespace.
+func (t *JenkinsPipelineTemplate) Instantiate() error {
+	if len(t.Errors()) > 0 {
+		return fmt.Errorf("unable to instantiate Jenkins, processing jenkins template failed")
+	}
+	if !t.hasJenkinsService() {
+		err := fmt.Errorf("template %s/%s does not contain required service %q", t.Config.Namespace, t.Config.TemplateName, t.Config.ServiceName)
+		t.CreateErrors = append(t.CreateErrors, err)
+		return err
+	}
+	counter := 0
+	for _, item := range t.items {
+		var err error
+		if item.IsOrigin {
+			err = t.osClient.Post().Namespace(t.TargetNamespace).Resource(item.Resource).Body(item.RawJSON).Do().Error()
+		} else {
+			err = t.kubeClient.Post().Namespace(t.TargetNamespace).Resource(item.Resource).Body(item.RawJSON).Do().Error()
+		}
+		if err != nil {
+			t.CreateErrors = append(t.CreateErrors, fmt.Errorf("creating Jenkins component %s/%s failed: %v", item.Kind, item.Name, err))
+			continue
+		}
+		counter++
+	}
+	delta := len(t.items) - counter
+	if delta != 0 {
+		// TODO: Shold we rollback in this case?
+		return fmt.Errorf("%d Jenkins pipeline components failed to create", delta)
+	}
+	return nil
+}
+
+// Errors returns the list of processing and creation errors.
+func (t *JenkinsPipelineTemplate) Errors() []error {
+	return append(t.ProcessErrors, t.CreateErrors...)
+}
+
+// resourceMapping specify resource metadata informations and JSON for items
+// contained in the Jenkins template.
+type resourceMapping struct {
+	Name     string
```

Comment:
@smarterclayton also the resource builder does not provide a way to have input different than stream, filename, stdin, etc... What I need is to feed []runtime.Object to it, otherwise I will have to convert the template items into JSON and then use IO reader... or maybe I'm missing something....